### PR TITLE
feat: Add mock mode for log collector with Tilt tests

### LINKED ISSUE #122: [Feature]: Support for automated log collection
### Prerequisites

- [x] I searched existing issues

### Feature Summary

NVSentinel automatically perform the recovery action for faults detected. In some cases, an operator will need to perform more deeper investigation on nodes that are failing repeatedly or nodes that are exhibiting new failure patterns. In order to do this, the operator need to have access to bug reports collected before the failure was remediated. 

### Problem/Use Case

Empower an operator with sufficient logs so that they can investigate failures beyond the initial recovery. These logs may also be used in the incident RCCA forums.

### Proposed Solution

Add a simple cluster local nginx based file server that can store the collected logs. The fault remediation module can spin up a job that will run the nvidia-bug-report, vendor provided SOS report and will upload it to the cluster local nginx server. 

### Component

Fault Management

### PR DESCRIPTION
**Pull request by rupalis-nv**
## Summary
This PR implements mock mode for the log collector to enable faster regression testing in Tilt environments without requiring actual GPU nodes, as specified in HIPPO-2120.

**Key Changes:**
- **Mock Mode Implementation**: Added auto-detection of KWOK fake nodes in `entrypoint.sh`. When detected, the log collector creates mock artifact files and uploads them to the file server, then exits with a configurable exit code.
- **Helm Configuration**: Added `mockMode`, `mockExitCode`, and `mockSleepDuration` parameters to the fault-remediation chart values, with `mockMode: auto` as the default for seamless Tilt integration.
- **Tilt Tests**: Implemented comprehensive Go-based tests (`TestLogCollectorSuccess` and `TestLogCollectorEndToEnd`) that verify:
  - Log collector job creation and completion
  - Mock artifact upload to file server
  - Full remediation flow including RebootNode CR creation
- **Extended Timeout**: Increased log collector job timeout from 5 to 10 minutes to accommodate larger clusters.

**Testing:**
- All tests pass successfully in Tilt environment
- Mock mode auto-detects KWOK nodes and creates/uploads artifacts correctly
- Tests are stable and non-flaky with timestamp filtering

## Type of Change
- [ ] 🐛 Bug fix
- [x] ✨ New feature
- [ ] 💥 Breaking change
- [ ] 📚 Documentation
- [ ] 🔧 Refactoring
- [ ] 🔨 Build/CI

## Component(s) Affected
- [ ] Core Services
- [ ] Documentation/CI
- [ ] Fault Management
- [ ] Health Monitors
- [ ] Janitor
- [x] Other: ____________

## Testing
- [x] Tests pass locally
- [x] Manual testing completed
- [x] No breaking changes (or documented)

## Checklist
- [ ] Self-review completed
- [ ] Documentation updated (if needed)
- [ ] Ready for review

## Summary by CodeRabbit

* **New Features**
  - Configurable mock mode for the log collector to simulate artifact creation, control exit code and sleep duration, and include mock tooling for testing.

* **Improvements**
  - Increased log collector job watch timeout from 5 to 10 minutes.
  - Entrypoint now supports mock-driven early exit and logs mock activity.

* **Configuration**
  - Added logCollector settings: mockMode, mockExitCode, mockSleepDuration and enabled in dev tilt config.

* **Tests**
  - New e2e tests and helpers for log-collector creation, completion, failure handling, artifact validation, and node-level overrides.

### CODE REVIEW

File: tests/log_collector_test.go [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+
+			for _, job := range jobs.Items {
+				if strings.Contains(job.GetName(), "log-collector") && job.CreationTimestamp.Time.After(testStartTime) {
+					logCollectorJob = &job
+					return true
+				}
+			}
+			return false
+		}, helpers.EventuallyWaitTimeout, helpers.WaitInterval, "log collector job should be created")
+		t.Logf("Log collector job created: %s", logCollectorJob.GetName())
```
Comment (tanishagoyal2) [inline tests/log_collector_test.go]: You can create log_collector_helper file and can move this code to there.

File: tests/log_collector_test.go [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+		require.Eventually(t, func() bool {
+			job := &batchv1.Job{}
+			err := client.Resources(helpers.NVSentinelNamespace).Get(ctx, logCollectorJob.GetName(), helpers.NVSentinelNamespace, job)
+			if err != nil {
+				return false
+			}
+			return job.Status.Succeeded > 0
+		}, helpers.EventuallyWaitTimeout, helpers.WaitInterval, "log collector job should complete")
+		t.Log(" Log collector job completed")
+
```
Comment (tanishagoyal2) [inline tests/log_collector_test.go]: same comment as above, please move repeated code to helper file

File: tests/log_collector_test.go [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+		// Step 4: Verify reboot CR is created
+		t.Log("Step 4: Waiting for RebootNode CR creation")
+		cr := helpers.WaitForRebootNodeCR(ctx, t, client, testCtx.NodeName)
+		t.Logf("RebootNode CR created: %s", cr.GetName())
+
+		// Step 5: Validate that mock artifacts were uploaded to file server
+		t.Log("Step 5: Validating mock artifacts were uploaded to file server")
```
Comment (tanishagoyal2) [inline tests/log_collector_test.go]: this also can move to common place

File: tests/log_collector_test.go [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+		require.NoError(t, err)
+
+		testStartTime := time.Now() // Record start time to filter old jobs
+
+		// Step 1: Trigger fault remediation
+		t.Log("Step 1: Triggering fault remediation flow")
+		helpers.TriggerFullRemediationFlow(ctx, t, client, testCtx.NodeName, 2)
+		t.Log("Node cordoned successfully")
+
+		// Step 2: Verify log collector job is created
```
Comment (tanishagoyal2) [inline tests/log_collector_test.go]: please remove obvious comments. this step intention is clear from log

**Review by tanishagoyal2 — CHANGES_REQUESTED** [review]

File: distros/kubernetes/nvsentinel/charts/fault-remediation/files/log-collector-job.yaml [hunk 5]
```diff
@@ -61,6 +61,12 @@ spec:
               value: {{ .Values.logCollector.enableGcpSosCollection | quote }}
             - name: ENABLE_AWS_SOS_COLLECTION
               value: {{ .Values.logCollector.enableAwsSosCollection | quote }}
+            - name: MOCK_MODE
+              value: {{ .Values.logCollector.mockMode | default "false" | quote }}
+            - name: MOCK_EXIT_CODE
+              value: {{ .Values.logCollector.mockExitCode | default "0" | quote }}
+            - name: MOCK_SLEEP_DURATION
+              value: {{ .Values.logCollector.mockSleepDuration | default "5" | quote }}
```
Comment (lalitadithya) [inline distros/kubernetes/nvsentinel/charts/fault-remediation/files/log-collector-job.yaml]: We don't have to add default here and in the values, let's keep it in the values file

File: log-collector/entrypoint.sh [hunk 6]
```diff
@@ -30,6 +30,64 @@ MUST_GATHER_SCRIPT_URL="${MUST_GATHER_SCRIPT_URL:-https://raw.githubusercontent.
...
+# Auto-detect KWOK nodes for mock mode
+if [ "${MOCK_MODE}" = "auto" ]; then
+    # Check if this is a KWOK fake node
+    if kubectl get node "${NODE_NAME}" -o jsonpath='{.metadata.annotations.kwok\.x-k8s\.io/node}' 2>/dev/null | grep -q "fake"; then
+        MOCK_MODE="true"
+        echo "[INFO] Auto-detected KWOK fake node - enabling mock mode"
+    else
+        MOCK_MODE="false"
+    fi
+fi
```
Comment (lalitadithya) [inline log-collector/entrypoint.sh]: On second thought since this need kube permissions, let's drop auto and keep only true/false

File: log-collector/entrypoint.sh [hunk 7]
```diff
@@ -30,6 +30,64 @@ MUST_GATHER_SCRIPT_URL="${MUST_GATHER_SCRIPT_URL:-https://raw.githubusercontent.
...
+
+# If mock mode is enabled, simulate log collection and exit
+if [ "${MOCK_MODE}" = "true" ]; then
```
Comment (lalitadithya) [inline log-collector/entrypoint.sh:42]: This changes the complete flow -- we are testing a different flow than what we are running prod. Can we instead make targeted checks to not run the actual command? Or even better would be if we can create simple fake versions of the tools somehow for mocks

File: tests/log_collector_test.go [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+	"github.com/stretchr/testify/require"
+	batchv1 "k8s.io/api/batch/v1"
+	"sigs.k8s.io/e2e-framework/pkg/envconf"
+	"sigs.k8s.io/e2e-framework/pkg/features"
+
+	"tests/helpers"
+)
+
+// TestLogCollectorSuccess tests that log collector job completes successfully in mock mode
+func TestLogCollectorSuccess(t *testing.T) {
```
Comment (lalitadithya) [inline tests/log_collector_test.go]: Can we test this as a part of the scale/smoke test? We can enable the flag in the values and we can do it as a part of a single flow

File: tests/log_collector_test.go [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+		require.Eventually(t, func() bool {
+			job := &batchv1.Job{}
+			err := client.Resources(helpers.NVSentinelNamespace).Get(ctx, logCollectorJob.GetName(), helpers.NVSentinelNamespace, job)
+			if err != nil {
+				return false
+			}
+			return job.Status.Failed > 0
+		}, helpers.EventuallyWaitTimeout, helpers.WaitInterval, "log collector job should fail")
+
+		t.Logf(" Log collector job failed as expected")
```
Comment (lalitadithya) [inline tests/log_collector_test.go]: Can we also check if the reboot node was created? Since the remediation should work even if the log collection fails.

File: tests/log_collector_test.go [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+
+	feature.Teardown(func(ctx context.Context, t *testing.T, c *envconf.Config) context.Context {
+		return helpers.TeardownFaultRemediation(ctx, t, c)
+	})
+
+	testEnv.Test(t, feature.Feature())
+}
+
+// TestLogCollectorEndToEnd tests the complete end-to-end flow with log collector
+func TestLogCollectorEndToEnd(t *testing.T) {
```
Comment (lalitadithya) [inline tests/log_collector_test.go]: This seems to overlap with the other two test cases, can you check? Can we have a small number of clean targeted tests?

File: tests/log_collector_test.go [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+		t.Log("Step 4: Waiting for RebootNode CR creation")
+		cr := helpers.WaitForRebootNodeCR(ctx, t, client, testCtx.NodeName)
+		t.Logf("RebootNode CR created: %s", cr.GetName())
+
+		// Step 5: Validate that mock artifacts were uploaded to file server
+		t.Log("Step 5: Validating mock artifacts were uploaded to file server")
+		fileServerPodName, err := helpers.GetFileServerPodName(ctx, client)
+		if err == nil && fileServerPodName != "" {
+			nodeNameStr := testCtx.NodeName
+			cmd := []string{"sh", "-c", "ls -la /data/" + nodeNameStr + "/ 2>/dev/null | grep -E 'nvidia-bug-report|gpu-operator-must-gather' || echo 'no files'"}
```
Comment (lalitadithya) [inline tests/log_collector_test.go]: Can we check the content of the files to make sure empty files are not being uploaded?

File: log-collector/entrypoint.sh (hunk 7 above)
Comment (rupalis-nv) [reply log-collector/entrypoint.sh:42]: Good catch
Can you help me with what is better approach to implement-
1. Function Overrides 
````
if [ "${MOCK_MODE}" = "true" ]; then
    run_nvidia_bug_report() {
        echo "[MOCK] Generating nvidia-bug-report"
        echo "Mock NVIDIA Bug Report - Node: ${NODE_NAME}, Timestamp: ${TIMESTAMP}" > "$1"
        return 0
    }
    run_must_gather() {
        echo "[MOCK] Running must-gather"
        tar czf "$1" -C /tmp --files-from=/dev/null
        return 0
    }
fi
run_nvidia_bug_report "${ARTIFACTS_DIR}/nvidia-bug-report.log.gz"
run_must_gather "${ARTIFACTS_DIR}/must-gather.tar.gz"
upload_artifacts_to_file_server 
````

2. Create Fake Tool Scripts
Create separate mock scripts (mock-nvidia-bug-report.sh, mock-must-gather.sh) and use PATH manipulation
In this we have to maintain mock scripts

File: tests/log_collector_test.go (hunk 8 above)
Comment (rupalis-nv) [reply tests/log_collector_test.go]: Should I delete the separate log_collector_test.go file entirely and only test log collector as part of smoke test?
For the mock mode flag - should it be enabled in the smoke test's values, or only in Tilt's values-tilt.yaml?

File: tests/log_collector_test.go (hunk 10 above)
Comment (rupalis-nv) [reply tests/log_collector_test.go]: So which approach is better?
1. Keep 2 tests 
TestLogCollectorSuccess - Full flow: Job success + File upload + RebootNode CR creation
TestLogCollectorFailure - Failure case: Job fails + RebootNode CR still created
2. Single test in smoke test
Delete all three separate tests
Add log collector verification to existing TestFatalHealthEvent smoke test

File: tests/log_collector_test.go (hunk 8 above)
Comment (lalitadithya) [reply tests/log_collector_test.go]: > should it be enabled in the smoke test's values, or only in Tilt's values-tilt.yaml?

yes

> Should I delete the separate log_collector_test.go file entirely and only test log collector as part of smoke test?

can you keep the negative test there and positive test merge it with the smoke test

File: tests/log_collector_test.go (hunk 10 above)
Comment (lalitadithya) [reply tests/log_collector_test.go]: > So which approach is better?

Single test for happy path, one dedicated test case for failure

File: log-collector/entrypoint.sh (hunk 7 above)
Comment (lalitadithya) [reply log-collector/entrypoint.sh:42]: Can you try 2 first if that works let's go with it else 1

File: tests/log_collector_test.go (hunk 8 above)
Comment (rupalis-nv) [reply tests/log_collector_test.go]: Do we really need negative test?
Because- 
- The fault-remediation code already handles this - log collection is non-blocking
- We have unit tests in fault-remediation that mock job failures

Comment (lalitadithya) [reply tests/log_collector_test.go]: Yes, we need to have end to end tests for user facing scenarios, especially negative ones
